### PR TITLE
cuDF Python reorg

### PR DESCRIPTION
**Pull request by shwina**
* Introduce `cudf.core` package and mirror `pandas.core` as closely as possible.

* Rename `cudf.bindings` to `cudf._lib`. Introduce the following convention for importing/using bindings (example):

```python
import cudf._lib as libcudf

libcudf.copying.scatter(...)
```

* Move `dtype` related utilities to `dtypes.py`.

### CODE REVIEW

File: cpp/tests/types/types_test.cpp [hunk 1]
```diff
@@ -283,7 +283,7 @@ TEST_F(WrappersTestBool8, BooleanOperatorsBool8)
 TEST_F(WrappersTestBool8, CastArithmeticTest)
 {
     cudf::bool8 w1{42};
-    cudf::bool8 w2{static_cast<char>(-42)};
+    cudf::bool8 w2{-42};
```
Comment (kkraus14) [inline cpp/tests/types/types_test.cpp]: Why did you change this back? This will cause compilation issues on ARM.

File: python/cudf/cudf/_lib/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+from . import (
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/__init__.py]: We've avoided relative imports in general thus far but I'm guessing we need it here?

File: cpp/tests/types/types_test.cpp (hunk 1 above)
Comment (shwina) [reply cpp/tests/types/types_test.cpp]: Great catch. I don't think I touched this but it's likely I borked something when resolving a merge conflict

File: python/cudf/cudf/_lib/__init__.py (hunk 2 above)
Comment (shwina) [reply python/cudf/cudf/_lib/__init__.py]: Yes - I gave it a lot of thought and it's the cleanest way to make all the modules available as attributes of the `_lib` package.

File: python/cudf/cudf/_lib/groupby.pyx [hunk 3]
```diff
@@ -69,8 +51,7 @@ def apply_groupby(
     values : list of Columns
     ops : str or list of str
         Aggregation to be performed for each column in *values*
-    dropna : bool
```
Comment (shwina) [inline python/cudf/cudf/_lib/groupby.pyx]: These changes were undone.

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,21 +4,20 @@
 
 - PR #2522 Add Java bindings for NVStrings backed upper and lower case mutators
 - PR #2607 Add Java bindings for parsing JSON
-- PR #2629 Add dropna= parameter to groupby
```
Comment (shwina) [inline CHANGELOG.md]: Need to un-undo this

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,21 +4,20 @@
...
 - PR #2571 Remove unnecessary managed memory from gdf_column_concat
+- PR #2648 Cython/Python reorg
 
 ## Bug Fixes
 
 - PR #2584 ORC Reader: fix parsing of `DECIMAL` index positions
 - PR #2614 Update Java version to match
 - PR #2601 Fixes nlargest(1) issue in Series and Dataframe
 - PR #2610 Fix a bug in index serialization (properly pass DeviceNDArray)
-- PR #2611 Types Test: fix static casting from negative int to string
```
Comment (shwina) [inline CHANGELOG.md]: Need to un-undo this

File: python/cudf/cudf/_lib/csv.pyx [hunk 6]
```diff
@@ -5,27 +5,32 @@
...
 import numpy as np
 import collections.abc as abc
 import os
 import errno
 
+from cudf._lib.includes.csv cimport (
+    reader as csv_reader,
+    reader_options as csv_reader_options,
+    csv_write_arg,
+    write_csv as cpp_write_csv
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/csv.pyx]: Any reason you didn't preface these with `cpp_`?

File: python/cudf/cudf/_lib/includes/copying.pxd [hunk 7]
```diff
@@ -5,29 +5,31 @@
...
 
 cdef extern from "cudf/copying.hpp" namespace "cudf" nogil:
 
-    cdef void gather(
-        const cudf_table * source_table,
+    void gather(
+        cudf_table * source_table,
         const gdf_index_type* gather_map,
         cudf_table* destination_table
     ) except +
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/includes/copying.pxd]: Why did we remove the cdef / const here?

File: python/cudf/cudf/_lib/join.pyx [hunk 8]
```diff
@@ -6,16 +6,15 @@
 # cython: language_level = 3
 
 # Copyright (c) 2018, NVIDIA CORPORATION.
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/join.pyx]: Duplicate copyright, add back in linebreak
```suggestion

```

File: python/cudf/cudf/_lib/csv.pyx (hunk 6 above)
Comment (shwina) [reply python/cudf/cudf/_lib/csv.pyx]: There weren't any naming conflicts that would be introduced by importing them as is. Could probably avoid this `cimport` altogether and just use `cpp_csv.reader` for example if you want consistency.

File: python/cudf/cudf/_lib/includes/copying.pxd (hunk 7 above)
Comment (shwina) [reply python/cudf/cudf/_lib/includes/copying.pxd]: My bad - will put them back in.

File: python/cudf/cudf/_lib/csv.pyx (hunk 6 above)
Comment (kkraus14) [reply python/cudf/cudf/_lib/csv.pyx]: I think that would be cleaner, what do you think?

Comment (shwina) [reply python/cudf/cudf/_lib/csv.pyx]: Done
